### ISSUE TITLE
Problem in Adding char*

### ISSUE THREAD

**Issue by srbehera** [efficient/cuckoofilter#2]
I am reading an input file which contains a string in each line. I am creating the filter like this
CuckooFilter<char*, 16> filter(total_items);
I can add items, but I cannot find the existing items.
It is working fine if I use static array e.g. char[31], but it does not work well for char \* or string.

**Comment by bingo4508** [comment]
The reason you cannot find the existing items is because you have to clear the char array to make sure the chars behind '\0' are the same.
ex:
char[31] str;
bzero(str, 31);

**Comment by srbehera** [comment]
It is still not working !
This is my code 
```
CuckooFilter<string, 12> filter(total_items);
string line;
while(getline(inp, line))
{
    if(filter.Contain(line) == cuckoofilter::Ok)
        cout<<"Found:"<<line<<endl;
    else
        filter.Add(cstr);
}

```

**Comment by bingo4508** [comment]
I think no char* and string now. It needs primative and non-pointer type because there is sizeof(item) in the code. So sizeof(string) will go wrong.

**Comment by apc999** [comment]
@bingo4508  you are right. currently the way to calculate hash index is [here ](https://github.com/efficient/cuckoofilter/blob/master/src/cuckoofilter.h#L65).
As a result, the string is converted into `char*`, getting into problems.

@srbehera for now string is not supported but only primary types (where `sizeof(item)` makes sense). I planed to support `string` but still not able to find time doing that :(

**Comment by srbehera** [comment]
@bingo4508 @apc999 Got it !! Thanks !!

**Comment by NoSeatbelts** [comment]
Would a templated function that defaulted to sizeof() for other structs/types with a specialization for std::string (using std::string.c_str() rather than (const char *) and std::string.size() rather than sizeof) be a quick fix?

**Comment by apc999** [comment]
yes, I think so.
Deaniel, feel free to submit a patch.
otherwise, I will probably add one later .

best,

**Comment by NoSeatbelts** [comment]
[Here](https://github.com/NoSeatbelts/cuckoofilter/tree/std_string)'s the patch I've made. It compiles -- I'm not sure if it fixes the problem, as I haven't built any test cases. Want to give it a whirl, srbehera?

**Comment by srbehera11** [comment]
Daniel, it is not working for string. 
Here is the simple code I was testing 

```
int main(int argc, char** argv) {
    CuckooFilter<string, 12> filter2(total_items);
    srand (time(NULL));
    ifstream inp;
    inp.open("test.txt");
    string line;
    int c1 = 0;
    int i = 0;
    while(getline(inp, line))
    {
      cout<<line<<endl;
      if(filter2.Contain(line) == cuckoofilter::Ok){
        cout<<"Found:"<<line<<endl;
        c++;
      }
      else{
        filter2.Add(line);
      }
      i++;
    }
    cout<< "c = "<< c <<endl;
    cout<< "i = "<< i <<endl;
    inp.close();
```
Where test.txt has following entries
```
AAA
BBB
CCC
AAA
CCC
DDD
XYZ
CCC
XYZ
AAA
```

**Comment by NoSeatbelts** [comment]
Is there any output? What does it do?

**Comment by srbehera11** [comment]
The program was putting the strings in the filter. If it is already there, it was printing "Found". The following output shows that it only puts the first string and then says "Found" all other strings.
```
AAA
BBB
Found:BBB
CCC
Found:CCC
AAA
Found:AAA
CCC
Found:CCC
DDD
Found:DDD
XYZ
Found:XYZ
CCC
Found:CCC
XYZ
Found:XYZ
AAA
Found:AAA
c1=9
i=10
```

**Comment by siara-cc** [comment]
I faced the same issue and it looks like according to the implementation the first parameter can only be `uint64_t`.
So I used following code:
```
std::string mystring = "Hello World";
cuckoofilter::CuckooFilter<size_t, 12> test(1000);
test.Add(CityHash64(mystring.c_str(), mystring.length()));
```
`CitiHash64` can be found in `src/City.h` in https://github.com/aappleby/smhasher
